[LoopIdiomRecognizer] Implement CRC recognition

joe-img opened the pull request:
Recognizes CRC byte loops and replaces them with a table lookup.

Current limitations:
- Only works on byte loops
	CRC size can be any, but the data is limited to one byte. i.e. a loop with iteration count 8.
- Only works on single-block loops
	most CRC loops would have been flattened to one block with select instructions this far into the pipeline. 

Both limitations were  in effort to reduce complexity, especially for a first patch. The code can be fairly easily extended to overcome these limitations.

Implementation details:
1) Check if the loop looks like CRC and extract some useful information
2) Execute one iteration of the instruction of the loop to see what happens to our output value
4) Ensure the output value is predicated on the value of the M/LSB of our input
4) Construct an expected output value of one iteration of CRC using the extracted information from step one and compare
5) Construct a lookup table and replace the output value with a lookup

joe-img committed:
[LoopIdiomRecognize] Implement function to extract CRC data from loops

This will check to see if a loop looks like CRC, not necessarily
guaranteeing that it is CRC.

joe-img committed:
[LoopIdiomRecognize] Implement ValueBits class

This is a representation of a value's bits in terms of references to
other values' bits, or 1/0 if the bit is known. This allows symbolic
execution of bitwise instructions without knowing the exact values.

Example:

LLVM IR Value i8 %x:
[%x[7], %x[6], %x[5], %x[4], %x[3], %x[2], %x[1], %x[0]]

%shr = lshr i8 %x, 2
[ 0, 0, %x[7], %x[6], %x[5], %x[4], %x[3], %x[2]]

%shl = shl i8 %shr, 1
[ 0, %x[7], %x[6], %x[5], %x[4], %x[3], %x[2], 0]

%xor = xor i8 %shl, 0xb
[ 0, %x[7], %x[6], %x[5], %x[4]^1, %x[3], %x[2]^1, 1]

joe-img committed:
[LoopIdiomRecognize] Implement PredicatedValueBits

These would be representitive of select or phi instructions where the
bits would depend on an icmp.

joe-img committed:
[LoopIdiomRecognize] Add function to symbolically execute basic block

The result is a map between llvm Values and their bit representations as
ValueBits.

joe-img committed:
[LoopIdiomRecognize] Check result of symbolic execution matches CRC

joe-img committed:
[LoopIdiomRecognize] Write lookup table based CRC for one-byte data

joe-img committed:
[LoopIdiomRecognize] Add unit tests for CRC idiom recognizer

joe-img commented:
RFC link: https://discourse.llvm.org/t/rfc-crc-recognition-in-loopidiomrecognizer/76491

mgudim commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -2868,3 +2889,883 @@ bool LoopIdiomRecognize::recognizeShiftUntilZero() {
+  // Symbolically execute one iteration of the loop to populate a map of
+  // Value's to their ValueBits, aka a representation of their bits in terms of
+  // 1's, 0's and references to other values' bits. If these match pre-computed
+  // crc values, then we can say it's doing crc.
```
Why is executing one iteration enough?

joe-img replied:
Updated. Does it make sense?

mgudim commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -2868,3 +2889,883 @@ bool LoopIdiomRecognize::recognizeShiftUntilZero() {
+    return false;
+  }
+
+  // If there exists a Data input, ensure the check bit is crc^data.
```
this comment is too specific to coremark.  Can you make it more general?

joe-img replied:
Looking at many CRC implementations, I disagree that this is specific to CoreMark. The data is almost always xor'd into the crc. Half of these, coremark included, decide to xor inside the loop.

mgudim commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -2868,3 +2889,883 @@ bool LoopIdiomRecognize::recognizeShiftUntilZero() {
+  }
+  ValueBits *ICmpOp0Bits = Result->second;
+
+  // Now match the following cases
```
I don't understand these. Can you write C code instead for each CRC variation you're trying to detect?

joe-img committed:
Clarify some comments

mgudim commented:
@joe-img I am going to review this properly, but now I am pretty busy with other work. Please ping me in a 2 weeks, if I haven't reviewed before, sorry for the long wait.

joe-img commented:
> @joe-img I am going to review this properly, but now I am pretty busy with other work. Please ping me in a 2 weeks, if I haven't reviewed before, sorry for the long wait.

Ping : )

mgudim commented:
> > @joe-img I am going to review this properly, but now I am pretty busy with other work. Please ping me in a 2 weeks, if I haven't reviewed before, sorry for the long wait.
> 
> Ping : )

Thanks for the reminder. I'm in the process.

topperc commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -135,6 +137,9 @@ static cl::opt<bool> UseLIRCodeSizeHeurs(
              "with -Os/-Oz"),
     cl::init(true), cl::Hidden);
 
+static cl::opt<bool> CRCRecognize("recognize-crc", cl::desc("CRC RECOGNIZE"),
```
Description should not be in all caps

topperc commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -2868,3 +2889,888 @@ bool LoopIdiomRecognize::recognizeShiftUntilZero() {
+    static ValueBit *CreateOneBit() { return new ValueBit(BitType::ONE); }
+    static ValueBit *CreateZeroBit() { return new ValueBit(BitType::ZERO); }
+    static ValueBit *CreateRefBit(Value *Ref, uint64_t Offset) {
+      return new ValueBit(BitType::REF, std::make_pair(Ref, Offset));
```
Is this memory ever freed? A quick search doesn't show any calls to `delete` or `unique_ptr`

topperc commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -2868,3 +2889,888 @@ bool LoopIdiomRecognize::recognizeShiftUntilZero() {
+        if (Incoming != BB) {
+          if (IncomingBlock) {
+            LLVM_DEBUG(dbgs()
+                       << DEBUG_TYPE " CRCRegonize: Do not know how to"
```
Recognize*. This spelling error is repeated in many messages.

topperc commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -2868,3 +2889,888 @@ bool LoopIdiomRecognize::recognizeShiftUntilZero() {
+  // than we need to account for overflow) and copying the 64bit values across
+  // aligned correctly
+  uint64_t CRCNumBytes = CRCSize / 8;
+  char *CRCTableData = (char *)malloc(CRCNumBytes * 260);
```
Please use array `new` and `delete`.

And use uint8_t or int8_t instead of char. The signedness of char is platform dependent

topperc commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -2868,3 +2889,888 @@ bool LoopIdiomRecognize::recognizeShiftUntilZero() {
+    CRCOffset = Builder.CreateXor(CRCOffset, Data);
+  }
+
+  CRCOffset = Builder.CreateZExt(CRCOffset, Builder.getInt32Ty());
```
You should probably get the index type from DataLayout and not hardcode i32

topperc commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -2868,3 +2889,888 @@ bool LoopIdiomRecognize::recognizeShiftUntilZero() {
+  // (MSB): crc < 0
+  // (MSB): crc <= -1
+  // And decide whether the check is checking for existence of 1 or 0
+  bool checkZero = false;
```
Capitalize variable name

topperc commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -2868,3 +2889,888 @@ bool LoopIdiomRecognize::recognizeShiftUntilZero() {
+    return std::nullopt;
+
+  // Check iteration count is 8
+  const SCEV *TripCountSCEV =
```
It's less code to use ScalarEvolution::getSmallConstantTripCount, but we already have the backedge count so I don't know what's better.

joe-img committed:
Review changes that aren't memory

joe-img committed:
Fix memory issues

- Use shared pointers for factory methods

I think the use of shared pointers in `symbolicallyExecute` is the
correct decision due to the filling of a map. I tried for a while to
make this a map of objects, but due to [object
slicing](https://en.wikipedia.org/wiki/Object_slicing) it would not be
possible to store a PredicatedValueBits in this map. (This would require
PredicatedValueBits to be redesigned into ValueBits class, but then the
internal methods get messy. Virtual functions was a clean answer to
this) Additionally, using `unique_ptr` proves difficult due to having to
store it in a map. And unique pointers cannot be copied.  As
`symbolicallyExecute` is the primary user of the factory methods, it
made sense for the factory methods to return a shared pointer instead of
a raw pointer or object and then wrapping this in a shared pointer
within symbolicallyExecute.

- Have XOR bits own their LHS/RHS

Having them as member objects instead of pointers I don't think makes
much sense due to them not being needed for most ValueBit objects, but
they would have to be initialized for all ValueBit instantiations.

- Use object values instead of pointers for bit representation

IMO the cleanest way. They're very lightweight anyway.

mgudim commented on llvm/test/Transforms/LoopIdiom/crc/crc.ll:
```diff
@@ -0,0 +1,195 @@
+define dso_local zeroext i8 @crc8_loop(ptr noundef %data, i32 noundef %length) {
+entry:
+  br label %for.cond
+
```
Please minimize the tests so that they only show the loop involved without all the extra code.

mgudim commented:
We should think through the high-level plan of how the implementation should work eventually for targets supporting this. At the same time, we can divide this into smaller pieces which has many advantages. 

I can suggest the following, but I really would appreciate feedback / approval of others (@kparzysz):

  (1) Create new (overloaded) intrinsics for performing a number of steps in polynomial division. Something like:
 `pdivstep(i8 dividend, i8 divisor, i32 nsteps)`. This intrinsic returns the result of performing `nsteps` steps in the polynomial divison of `dividend` by the `divisor`. Here it is assumed that the highest degree of polynomial corresponds to the MSB of the number representing it. Actual loop could be using reverse representation, but we can express it by combining with bitreverse intrinsics.

  (2) Add functions to `PreISelIntrinsicLowering` to expand the pdivstep intrinsic. We can expand back into a loop (or unrolled loop) like `expandMemCpyAsLoop`. Or we can emit a table lookup (like in this MR), or we can expand using `clmul` intrinsics.
For testing purposes, we can introduce flags to always (if possible) do specified type of expansion. We can start with "expand into a loop" and a "table lookup". We can add expansion using `clmul`s later.
  
  (3) Implement the recognition part in `LoopIdiomRecognize`. As in this MR, the generation of the intrinsic should be controlled by a flag. Also, the idiom should be "polynomial division", CRC is just one application of it. I don't know any other though :-)  If the whole plan is approved, we can merge this part first.

  (4) Recognize identities on `pdiv` intrinsics. I think in coremark, the 8-bit ones combine eventually into larger one. This should be after inlining, so probably we can do it in `PreISelIntrinsicLowering`?

  (5) Create a way for targets to specify if `pdiv` intrinsics should be generated. If yes, how should they be expanded (during `PreISelLowering`)?

joe-img commented:
Hi both and thanks for the comments. Sorry, I am working on this PR, I've just had some higher priority items.

> pdivstep(i8 dividend, i8 divisor, i32 nsteps). This intrinsic returns the result of performing nsteps steps in the polynomial divison of dividend by the divisor

So the return value is the quotient? Doesn't that complicate the expression of CRC? What we need for CRC is surely something like `premstep` instead?

> Actual loop could be using reverse representation, but we can express it by combining with bitreverse intrinsics.

Would it be easier to have a bit reversed variation of the intrinsic? We'd have to identify these bitreverse intrinsics upon expansion anyway. It would just avoid having to insert and then recognize it.


In general I think this is a good plan, and was the original vision. If this whole plan is approved, would the first MR be just recognition and creation of the division intrinsic? I.E strip the table based lookup from this MR and emit intrinsic but would error in the backend.

mgudim commented:
> So the return value is the quotient? Doesn't that complicate the expression of CRC? What we need for CRC is surely something like premstep instead?

Yes, you're right.

> Would it be easier to have a bit reversed variation of the intrinsic? We'd have to identify these bitreverse intrinsics upon expansion anyway. It would just avoid having to insert and then recognize it.

I am OK with that too. I just didn't want to introduce more intrinsics than we need.

>  would the first MR be just recognition and creation of the division intrinsic? 

I think first MR should be just defining the intrinsics.

> would the first MR be just recognition and creation of the division intrinsic? I.E strip the table based lookup from this MR 

yes.

> but would error in the backend.

Only emit intrinsic if a flag is passed so it doesn't break anything. Name of the flag could indicate that at this point the code gen doesn't work yet for the intrinsic. But at the same time you should be able to add tests to your MR.

mgudim commented:
> So the return value is the quotient? Doesn't that complicate the expression of CRC? What we need for CRC is surely something like premstep instead?

Yes, you're right. Actually, maybe we need both? In coremark after we emit crc8 intinsic, after inlining they combine to a larger crc (32 I think). To recognize that, we need to see that after doing 8-bit crc we continue with the quotient that we got, right? My memory is vague on this, so please correct me if I am wrong.

Sorry for all the noise. Looks like we only need the remainder after all. The identity is something like:

```
prem((prem(x, d, k_1) << s) || y), d, k_2) = prem((x <<s ) || y, d, k_1 + k_2
```

joe-img commented:
> > So the return value is the quotient? Doesn't that complicate the expression of CRC? What we need for CRC is surely something like premstep instead?
> 
> Yes, you're right. Actually, maybe we need both? In coremark after we emit crc8 intinsic, after inlining they combine to a larger crc (32 I think). To recognize that, we need to see that after doing 8-bit crc we continue with the quotient that we got, right? My memory is vague on this, so please correct me if I am wrong.
> 
> Sorry for all the noise. Looks like we only need the remainder after all. The identity is something like:
> 
> ```
> prem((prem(x, d, k_1) << s) || y), d, k_2) = prem((x << s ) || y, d, k_1 + k_2)
> ```

I'm a bit confused with all those arguments to prem. What do they represent?

Also, what is considered a 'step' in polynomial division?

In CoreMark, the helper function `ee_u16 crcu8(ee_u8 data, ee_u16 crc )` is used, but this is still crc16, as the generator polynomial and remainder are 16-bits wide. I think it's called 'crcu8' because it xor's in 8 bits of data.

If we're using a prem intrinsic, then surely it makes most sense for it to be something like `prem(divisor, dividend)`, where it returns the remainder after polynomial division? With this, I see two issues:
Firstly, for crcn, prem would have to take a divisor of bitsize 2n. CRC implementations avoid this by using a linear feedback shift register.

Secondly, I think this complicates the expression of crc. For example, I'm pretty sure that coremark's crcu8 will be represented like:

`(crc << 8) ^ prem((((crc & 0xff00) >> 8) ^ (data)) << 16), poly | (1 << 16))`

Although crc is the remainder of polynomial division accross some stream of data, the way it's implemented in crc, I don't think matches up very well with a polynomial remainder intrinsic.

> We can expand back into a loop (or unrolled loop) like expandMemCpyAsLoop

Is this necessary? If in LoopIdiomRecognize we're recognizing the loop, can we just avoid emitting the intrinsic using the same logic as would be in  `PreISelIntrinsicLowering` to decide whether to lower to loop?

zqb-all commented on llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp:
```diff
@@ -2868,3 +2893,928 @@ bool LoopIdiomRecognize::recognizeShiftUntilZero() {
+        OS << "0";
+        break;
+      case BitType::REF:
+        OS << _BitRef.first->getNameOrAsOperand() << "[" << _BitRef.second
```
`getNameOrAsOperand` causes a build error when building a release version

joe-img replied:
I've rebased & pushed, but don't see this issue. Can you recreate it now?

zqb-all replied:
I looked at some other code that uses `getNameOrAsOperand`, either wrapped in the `#ifndef NDEBUG` macro, or in `LLVM_DEBUG`

https://github.com/llvm/llvm-project/blob/919187871d142496d53bca8732677e0811b32409/llvm/include/llvm/IR/Value.h#L293-L295

joe-img replied:
You're dead right. Thanks for pointing this out.

joe-img committed:
Format code correctly

joe-img committed:
Surround ValueBit(s) printing with NDEBUG

getNameOrAsOperand is only available in debug mode, so causes failures
when built in release.

mgudim commented:
@joe-img I went through the code and just from looking this seems to work correctly. However, I still stand by my suggestions in this comment: https://github.com/llvm/llvm-project/pull/79295#issuecomment-1962269129 

I agree that the interpreter approach is better than trying to recognize all variations of the pattern in the source code.

I don't insist that you follow my suggestions and I am OK with someone else approving this as long as this is not enabled by default. This way we can merge your work and maybe improve on it gradually.

zqb-all commented on llvm/test/Transforms/LoopIdiom/crc/not-crc.ll:
```diff
@@ -0,0 +1,113 @@
+; RUN: opt -passes=loop-idiom < %s -S -debug -recognize-crc 2>&1 | FileCheck %s
+
+; crc16 incorrect xor inside loop
+; CHECK: loop-idiom CRCRegonize: Cannot verify check bit!
```
some `CRCRegonize `should fix to  `CRCRecognize `in this file

artagnon reviewed (commented):
While I think the idea of tracking known bits for CRC recognition is sound, I think there are problems with the implementation rolling its own version of KnownBits, creating an unnecessary PValueBits, and not using ConstantRange for the significant-bit check. The way it computes known bits using a Map is sub-optimal, and the patch pattern-matches too strongly. Besides, most of the work of optimizing a CRC is the analysis, and the transform bit is trivial. Also, this can be generalized to polynomial hashes in Galois field of characteristic 2, creating a rationale for a new analysis. I've taken the basic idea of tracking known bits, and implemented something in ~500 lines here: https://github.com/llvm/llvm-project/pull/139120.

artagnon commented:
CRC optimization has now landed, due to another patch. Closing this stale PR.
